core/migration.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 Solus Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"fmt"
+	"github.com/BurntSushi/toml"
+	log "github.com/DataDrake/waterlog"
+	"io/ioutil"
+	gouser "os/user"
+	"strconv"
+)
+
+// Migration contains the information about a migration, including where it is on the system and its requested system modifications
+type Migration struct {
+	Name string
+	Path string
+
+	Description string         `toml:"description"`
+	UpdateUsers []*UpdateUsers `toml:"users-update"`
+	UpdateGroup []*UpdateGroup `toml:"group-update"`
+}
+
+// UpdateUsers is a type of modification that adds a group to a specific set of users
+type UpdateUsers struct {
+	UserFilters []string `toml:"only"`
+	GroupName   string   `toml:"group"`
+}
+
+// UpdateGroup is a type of modification that attempts to create or update a group to the given name and ID
+type UpdateGroup struct {
+	GroupName  string `toml:"name"`
+	NewGroupID int    `toml:"id"`
+}
+
+// LoadMigrations finds migration files in SysDir and UsrDir and attempts to load them
+func LoadMigrations() []Migration {
+	var allMigrations []Migration
+
+	if sysFiles, err := ioutil.ReadDir(SysDir); err != nil {
+		log.Warnf("System directory for migrations at %s is unreadable, skipping\n", SysDir)
+	} else {
+		log.Debugf("Loading migrations from system directory %s...\n", SysDir)
+		for _, it := range sysFiles {
+			allMigrations = appendMigrationFrom(allMigrations, SysDir, it.Name())
+		}
+	}
+
+	if usrFiles, err := ioutil.ReadDir(UsrDir); err != nil {
+		log.Warnf("User directory for migrations at %s is unreadable, skipping\n", UsrDir)
+	} else {
+		log.Debugf("Loading migrations from user directory %s...\n", UsrDir)
+		for _, it := range usrFiles {
+			allMigrations = appendMigrationFrom(allMigrations, UsrDir, it.Name())
+		}
+	}
+
+	return allMigrations
+}
+
+func appendMigrationFrom(migrations []Migration, dir string, name string) []Migration {
+	if migration, err := parseMigration(dir, name); err != nil {
+		log.Warnf("\tFailed to parse migration %s: %s\n", name, err)
+	} else {
+		migrations = append(migrations, *migration)
+		log.Debugf("\tLoaded migration %s\n", name)
+	}
+	return migrations
+}
+
+func parseMigration(dir string, name string) (migration *Migration, err error) {
+	// Read the configuration into the program
+	path := fmt.Sprintf("%s/%s", dir, name)
```

Comment:
Should be using `filepath.Join` instead.

https://golang.org/pkg/path/filepath/#Join